Source/DafnyLanguageServer/Language/SemanticTokens/DafnySemanticTokensBuilder.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using Microsoft.Boogie;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.Language.SemanticTokens;
+
+public class DafnySemanticTokensBuilder {
+  private readonly SemanticTokensBuilder builder;
+
+  public DafnySemanticTokensBuilder(SemanticTokensBuilder builder) {
+    this.builder = builder;
+  }
+
+  public void Push(string source, IToken? tok, SemanticTokenType tokenType, params SemanticTokenModifier[] modifiers) {
+    if (tok == null || tok == Token.NoToken) {
+      Console.Error.WriteLine($"{source}: null");
+      return;
+    }
+    var pos = tok.GetLspPosition();
+    Console.Error.WriteLine($"{source}: ({pos.Line}, {pos.Character}) {tok.kind} = '{tok.val}'");
+    builder.Push(pos.Line, pos.Character, tok.val.Length, tokenType, modifiers);
+  }
+
+  public void Push(string source, IToken? tok, IToken? endTok, SemanticTokenType tokenType, params SemanticTokenModifier[] modifiers) {
+    if (tok == null || tok == Token.NoToken) {
+      Console.Error.WriteLine($"{source}: null");
+      return;
+    }
+    if (endTok == null || endTok == Token.NoToken) {
+      Push(source, tok, tokenType, modifiers);
+      return;
+    }
+    var range = new Range(tok.GetLspPosition(), endTok.GetLspPosition());
+    Console.Error.WriteLine($"{source}: ({range.Start.Line}, {range.Start.Character})-({range.End.Line}, {range.End.Character}) {tok.kind} = '{tok.val}'");
+    builder.Push(range, tokenType, modifiers);
+  }
+
+  public void Push(string source, IToken? tok) {
+    if (tok == null || tok == Token.NoToken) {
+      Console.Error.WriteLine($"{source}: null");
+      return;
+    }
+
+    var noModifiers = Array.Empty<SemanticTokenModifier>();
+    (SemanticTokenType? tokenType, SemanticTokenModifier[] modifiers) =
+#pragma warning disable CS8524
+      // [CS8524] The switch expression does not handle some values of its input type (it is not exhaustive) involving
+      // an unnamed enum value. For example, the pattern '(Microsoft.Dafny.Parser.TokenClass)16' is not covered.
+      Parser.TokenClassificationMap[tok.kind] switch {
```

Comment:
There are two levels of abstraction here. TokenClass regroups tokens by more classes than SemanticTokenType, so that it will be easier to migrate groups to new semantics tokens when they appear.